persistent/persistence.py

```diff
@@ -333,29 +342,54 @@ def _p_activate(self):
         """
         before = self.__flags
         if self.__flags is None or self._p_state < 0: # Only do this if we're a ghost
-            self.__flags = 0
+            _OSA(self, '_Persistent__flags', 0)
             if self.__jar is not None and self.__oid is not None:
                 try:
                     self.__jar.setstate(self)
                 except:
-                    self.__flags = before
+                    _OSA(self, '_Persistent__flags', before)
                     raise
 
+    # In the C implementation, _p_invalidate winds up calling
+    # _p_deactivate. There are ZODB tests that depend on this;
+    # it's not documented but there may be code in the wild
+    # that does as well
+
     def _p_deactivate(self):
         """ See IPersistent.
         """
         if self.__flags is not None and not self.__flags:
-            self._p_invalidate()
+            self._p_invalidate_deactivate_helper()
 
     def _p_invalidate(self):
         """ See IPersistent.
         """
+        # If we think we have changes, we must pretend
+        # like we don't so that deactivate does its job
+        _OSA(self, '_Persistent__flags', 0)
+        self._p_deactivate()
+
+    def _p_invalidate_deactivate_helper(self):
         if self.__jar is not None:
             if self.__flags is not None:
-                self.__flags = None
+                _OSA(self, '_Persistent__flags', None)
             idict = getattr(self, '__dict__', None)
             if idict is not None:
                 idict.clear()
+            # Implementation detail: deactivating/invalidating
+            # updates the size of the cache (if we have one)
+            # by telling it this object no longer takes any bytes
+            # (-1 is a magic number to compensate for the implementation,
+            # which always adds one to the size given)
+            try:
+                cache = self.__jar._cache
+            except AttributeError:
+                pass
+            else:
+                cache.update_object_size_estimation(self.__oid,
+                                                    -1)
```

Comment:
Ahh. It used to be a much longer line before I ran into the AttributeError and had to split it up. I'll condense it.